### ISSUE TITLE
http-client OSGi bundle doesn't allow instantiation of HonoClient

### ISSUE THREAD

**Issue by avgustinmm** [eclipse/hono#586]
AFAIK the only way to use HoneClient is to instantiate it via calling the direct constructor of org.eclipse.hono.client.impl.HonoClientImpl.
The org.eclipse.hono.client.impl, however, is not expored by the hono-client bundle. Only the org.eclipse.hono.client is exported.
So, the hono-client shall export org.eclipse.hono.client.impl (exporting impl, IMHO, is not recommended) or the org.eclipse.hono.client package shall provide meant to instantiate HonoClient without directly refer org.eclipse.hono.client.impl

**Comment by ctron** [comment]
I will look into this.

**Comment by sophokles73** [comment]
Maybe we should add a default factory method to the `HonoClient` interface, very much like vert.x does it all over the place ...

**Comment by avgustinmm** [comment]
Yes, this will solve the issue and will improve even non-OSGi usage. Then there will be no need of using impl classes directly - better API

**Comment by sophokles73** [comment]
I was more thinking about just adding something like the following to `HonoClient`

```
HonoClient createClient(Vertx vertx, ClientConfigProperties config) {
  return new HonoClientImpl(vertx, config);
}
```

I do not think that we need to change anything else for this, do we?

**Comment by ctron** [comment]
Well that would ignore the case of using an explicit `ConnectionFactory`.

**Comment by ctron** [comment]
I just pushed the (unfinished!) branch of the changes I had in mind.

**Comment by ctron** [comment]
Ok, pushed again without the code formatter changes.

**Comment by ctron** [comment]
The javadoc part and the refactoring of the "ConnectionBuilder" is still missing of course.

**Comment by ctron** [comment]
This should be fixed by #593
